// babel-plugin-transform-import-resolve/lib/index.js
'use strict';

var _slash = require('slash');

var _slash2 = _interopRequireDefault(_slash);

var _path = require('path');

var _resolve = require('resolve');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function parseResolve() {
  var resolved = _path.relative.apply(undefined, arguments);
  var prefixed = resolved.startsWith('.') || resolved.startsWith('/') ? resolved : './' + resolved;
  var forward = (0, _slash2.default)(prefixed);
  return forward;
}

function resolver(_ref, _ref2) {
  var source = _ref.node.source;
  var filename = _ref2.file.opts.filename;
  var _ref2$opts$parse = _ref2.opts.parse;
  var parse = _ref2$opts$parse === undefined ? parseResolve : _ref2$opts$parse;

  if (source !== null) {
    var basedir = (0, _path.dirname)(filename);
    var dependency = void 0;
    try {
      dependency = (0, _resolve.sync)(source.value, { basedir: basedir });
    } catch (err) {
      return;
    }
    source.value = parse(basedir, dependency, source.value, filename);
  }
}

function transformImportResolve() {
  return {
    visitor: {
      ExportNamedDeclaration: resolver,
      ImportDeclaration: resolver
    }
  };
}

module.exports = transformImportResolve;
